### ISSUE TITLE
Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4254]
This task is automatically imported from the old Task Issue Board and it was originally created by James Dunkerley.
Original issue is [here](https://www.pivotaltracker.com/story/show/181449213).
<hr />

From: https://github.com/enso-org/enso/issues/1600#issuecomment-1023420479

Working with atoms I've found out that if the default arguments are primitives, then they work except for the first argument (i.e. if you provide 0 args you get an AtomConstructor, but if you provide at least one, the others get defaulted correctly).

But for some reason if the default argument is itself a composite Atom, the interpreter crashes :confused: 

Best to illustrate it on this small example:
```
from Standard.Base import all

type Case_Insensitive
type Matching_Strategy
    type Exact (cs : (True | Case_Insensitive) = True)

type Column_Selector
    type By_Name (names : Vector Text) (ms : Matching_Strategy = Exact True)

# Interestingly, if there are multiple arguments but the default is a primitive, it seems to work:
type Multi bar=0 baz=42

main =
    y = Multi 23
    IO.println y

    z = Multi
    IO.println z

    w = Multi ["bar"]
    IO.println w
    
    # But this will break hard
    x = By_Name ["Foo"] # note the last argument is missing
    IO.println x
```

which produces the following output:

```
(Multi 23 42)
Constructor<Multi>
(Multi ['bar'] 42)
Execution finished with an error: java.lang.ClassCastException: class org.enso.interpreter.node.expression.builtin.InstantiateAtomNode cannot be cast to class org.enso.interpreter.node.EnsoRootNode (org.enso.interpreter.node.expression.builtin.InstantiateAtomNode and org.enso.interpreter.node.EnsoRootNode are in unnamed module of loader com.oracle.graalvm.locator.GraalVMLocator$GuestLangToolsLoader @7ef27d7f)
        at <java> org.enso.interpreter.node.BaseNode.getStateFrameSlot(BaseNode.java:35)
        at <java> org.enso.interpreter.node.callable.ApplicationNode.executeGeneric(ApplicationNode.java:97)
        at <java> org.enso.interpreter.node.callable.argument.ReadArgumentNode.executeGeneric(ReadArgumentNode.java:56)
        at <java> org.enso.interpreter.node.expression.atom.InstantiateNode.executeGeneric(InstantiateNode.java:60)
        at <java> org.enso.interpreter.node.expression.builtin.InstantiateAtomNode.execute(InstantiateAtomNode.java:33)
        at <java> org.enso.interpreter.node.expression.builtin.InstantiateAtomNode.execute(InstantiateAtomNode.java:12)
        at <enso> Main.main(src/Main.enso:24:9-23)
```

## Comments:
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-03-04):

**Today:** Started looking at the implementation of default arguments It should be finished by 2022-03-08.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181449213). Same task

**Additional Information:** Gave a rough estimate for Tuesday but it's a bit unclear at this stage to me how involving the bug will be. (Enso Bot - Mar 4, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-03-07):

**Today:** Debugging the issue It should be finished by 2022-03-08.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181449213). Same task

**Additional Information:** The End Date is likely to be delayed since I'm still digging into internals of the engine to find the root cause, and learning how things work on the way. This was expected. (Enso Bot - Mar 7, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-03-08):

**Summary:** There is **1 day** delay in implementation of the [Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors (#181449213) task](https://www.pivotaltracker.com/story/show/181449213).
It **will cause 1 day delay** for the delivery of this sprint.

Not sure whether it will be a day or 2, will likely submit an early draft PR to get some feedback about the solution

**Delay Cause:** I have a potential solution that replaces InstantiateAtomNode class with ClosureRootNode except that I discovered some scoping issues on the way. Might be a related or a new bug, though.

**Possible solutions:** Gain more experience with the engine 😉 (Enso Bot - Mar 9, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-08):

**Today:** After digging through the internals figured out more or less the reason for the failure. It should be finished by 2022-03-09.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181449213). Same task

**Additional Information:** The delay was kind of expected due to my unfamiliarity of the compiler. Need to find some time to setup intellij debugger with Enso. (Enso Bot - Mar 9, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-03-09):

**Summary:** There is **2 day** delay in implementation of the [Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors (#181449213) task](https://www.pivotaltracker.com/story/show/181449213).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** We extended the scope of the ticket to cover a new bug (Enso Bot - Mar 10, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-09):

**Today:** PR with the fix for the original crash is in, along with tests. But extending the scope due to another problem discovered on the way. It should be finished by 2022-03-11.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181449213). Same task (Enso Bot - Mar 10, 2022)
<hr />
Needs to cover a new case
```
from Standard.Base import all

type Foo a=True b=a
main =
    f = Foo False
    IO.println f
```
currently gives 
```
(Foo False UnresolvedSymbol<a>)
``` (Hubert Plociniczak - Mar 10, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-03-10):

**Today:** Still debugging. Found a quick hack for the new problem, need to generalize it so that it covers more complicated expressions as default args. It should be finished by 2022-03-11.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181449213). Same task

**Additional Information:** Slowly making progress in familiarizing myself with the engine. (Enso Bot - Mar 10, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for the provided date (2022-03-11):

**Summary:** There is **3 days** delay in implementation of the [Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors (#181449213) task](https://www.pivotaltracker.com/story/show/181449213).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** Overoptimistic end-date due to scope extension (Enso Bot - Mar 14, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for the provided date (2022-03-11):

**Today:** Hit a dead end, managed to hook up debugger, identified the problem but still working on a generic solution. It should be finished by 2022-03-12.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/181449213). Same task (Enso Bot - Mar 14, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-03-14):

**Summary:** There is **2 days** delay in implementation of the [Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors (#181449213) task](https://www.pivotaltracker.com/story/show/181449213).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** Wrong delay in the previous report (Enso Bot - Mar 15, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-03-14):

**Summary:** There is **3 days** delay in implementation of the [Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors (#181449213) task](https://www.pivotaltracker.com/story/show/181449213).
It **will cause 0 days delay** for the delivery of this sprint.

Trying to convince the bot that the delay was fine, sorry for the noise

**Delay Cause:** Wrong delay in the previous report (Enso Bot - Mar 15, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-14):

**Today:** Decided to rework the initial PR for the original issue, at the same time addressing more examples. Had to figure out a way to store values of already evaluated arguments so that they could be referenced later. Waiting on review of the draft solution. It should be finished by 2022-03-15.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). See the review and start looking at the new task. (Enso Bot - Mar 15, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-03-15):

**Summary:** There is **2 days** delay in implementation of the [Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors (#181449213) task](https://www.pivotaltracker.com/story/show/181449213).
It **will cause 0 days delay** for the delivery of this sprint.

**Delay Cause:** Scope expansion, addressing PR review (Enso Bot - Mar 16, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-15):

**Today:** Addressing PR review re performance. Have a working solution that mimics the solution used in functions. It should be finished by 2022-03-16.

**Next Day:** Next day I will be working on the [##181499077 task](https://www.pivotaltracker.com/story/show/181499077). Bringing the PR out of draft state, moving on to the next task. (Enso Bot - Mar 16, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for today (2022-03-16):

**Today:** Final PR ready for review It should be finished by 2022-03-16.

**Next Day:** Next day I will be working on the [##181497846 task](https://www.pivotaltracker.com/story/show/181497846). Addressing any further review but fairly confident of the state of it now

**Additional Information:** Took more time than I wished but learned a number of things on the way. (Enso Bot - Mar 16, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for the provided date (2022-03-15):

**Summary:** There is **1 day** delay in implementation of the [Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors (#181449213) task](https://www.pivotaltracker.com/story/show/181449213).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Scope expansion, addressing PR review (Enso Bot - Mar 18, 2022)
<hr />
**Hubert Plociniczak** reports a new **🔴 DELAY** for yesterday (2022-03-17):

**Summary:** There is **1 day** delay in implementation of the [Investigate and solve the "sub-bug" within Default Arguments in Atom Constructors (#181449213) task](https://www.pivotaltracker.com/story/show/181449213).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Scope expansion, addressing PR review (Enso Bot - Mar 18, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-03-17):

**Progress:** Addressing last feedback It should be finished by 2022-03-17.

**Next Day:** Next day I will be working on the [##181497846 task](https://www.pivotaltracker.com/story/show/181497846). Discovery & planning for the new task (Enso Bot - Mar 18, 2022)
<hr />
